File: 155.cpp
#include <stack>
using namespace std;

class MinStack {
  struct Datum {
    int val;
    int currMin;
  };
  stack<Datum> s;

public:
  MinStack() {}

  void push(int val) {
    if (s.empty()) {
      s.push({val, val});
    } else {
      s.push({val, min(val, s.top().currMin)});
    }
  }

  void pop() { s.pop(); }

  int top() { return s.top().val; }

  int getMin() { return s.top().currMin; }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 155.py
class MinStack:
    def __init__(self):
        self.s = []

    def push(self, val: int) -> None:
        # if the stack is empty
        if len(self.s) == 0:
            self.s.append((val, val))
        # if there are other elts in the stack
        else:
            self.s.append((val, min(val, self.s[-1][1])))

    def pop(self) -> None:
        self.s.pop()

    def top(self) -> int:
        return self.s[-1][0]

    def getMin(self) -> int:
        return self.s[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
